File: csharp/components/Strata.Persistence/InMemory/InMemoryNamedQuery.cs
//  ##########################################################################
//  # File Name: InMemoryNamedQuery.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using System;
using System.Collections.Generic;
using System.Linq;
using Strata.Foundation.Utility;
using Strata.Persistence.NamedQuery;

namespace Strata.Persistence.InMemory
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    ///  
    public 
    class InMemoryNamedQuery<T>:
        AbstractNamedQuery<T>
        where T:class
    {	
        private InMemoryUnitOfWorkProvider itsContext;
	    private Type                      itsType;
	    private IPredicate<T>			  itsPredicate;
	    private IList<T>				  result;
	    private IEnumerator<T>			  current;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// 
        /// </summary>
        /// 
        public 
        InMemoryNamedQuery(
            InMemoryUnitOfWorkProvider cntx,
            String                     name,
            IPredicate<T>              pred):
            base(name)
        {
            itsContext   = cntx;
		    itsType      = typeof(T);
		    itsPredicate = pred;
		    result   = null;
		    current  = null;
		
		    itsContext.InsertNamedQuery<T>(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// 
        /// </summary>
        /// 
        public   
        InMemoryNamedQuery(InMemoryNamedQuery<T> other):
            base( other )
        {
		    itsContext   = other.itsContext;
		    itsType      = other.itsType;
		    itsPredicate = other.itsPredicate;
		    result       = null;
		    current      = null;
        } 

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override ICopyable
        MakeCopy()
        {
            return new InMemoryNamedQuery<T>( this );            
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void 
        Clear()
        {
            ClearInputs();
            result  = null;
            current = null;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
	    public override void
	    Execute()
	    {
		    if ( result != null )
			    return;
		
		    result  = GetResults( 
                GetInputs().GetNamedInputs(),
                itsContext.GetEntitiesByType<T>() );
		    current = result.GetEnumerator();
	    }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override ICollection<T> 
        GetAll()
        {
		    Execute();
		    return result;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override T 
        GetUnique()
        {
		    Execute();

            return result.SingleOrDefault();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override T 
        GetNext()
        {
		    Execute();

            try
            {
                current.MoveNext();
                T next = current.Current;
                
                return next;
            }
            catch (InvalidOperationException)
            {
                return null;
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override bool 
        HasUnique()
        {
            Execute();
            return result.Count == 1;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override bool 
        HasAny()
        {
            Execute();
            return result.Count > 0;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override bool 
        HasNext()
        {
		    Execute();

            try
            {
                T next = current.Current;
                
                return true;
            }
            catch (InvalidOperationException)
            {
                return false;
            }
        }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// 
        /// </summary>
        /// 
	    protected IList<T> 
	    GetResults(IDictionary<String,Object> inputs,IList<T> objects)
	    {
		    IList<T> output = new List<T>();
		
		    foreach (T entry in objects)
			    if ( itsPredicate.Evaluate( entry,inputs ) )
				    output.Add( entry );
		
		    return output;
	    }

    }
}

//  ##########################################################################

File: legacy/java/components/StrataPersistence/src/strata/persistence/inmemory/InMemoryNamedQuery.java
// ##########################################################################
// # File Name:	InMemoryNamedQuery.java
// #
// # Copyright:	2011, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataEntity Framework.
// #
// #   			The StrataEntity Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataEntity Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataEntity
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.persistence.inmemory;



import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import strata.persistence.namedquery.AbstractNamedQuery;
import strata.persistence.namedquery.INamedQuery;
import strata.persistence.namedquery.NotUniqueException;

/****************************************************************************
 * 
 * @param <K> key type
 * @param <T> entity type
 * 
 * @author 		
 *     Sapientia Systems 
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class InMemoryNamedQuery<T> 
    extends    AbstractNamedQuery<T>
	implements INamedQuery<T>
{
	private InMemoryUnitOfWorkProvider itsProvider;
	private Class<T>                   itsType;
	private IPredicate<T>			   itsPredicate;
	private List<T>				       itsObjects;
	private List<T>				       itsResults;
	private Iterator<T>				   itsIterator;
		
	/************************************************************************
	 * Creates a new InMemoryNamedQuery. 
	 * @param type TODO
	 *
	 */
	public 
	InMemoryNamedQuery(
	    InMemoryUnitOfWorkProvider provider,
		String                     name,
		Class<T>                   type, 
		IPredicate<T>			   predicate)
	{
		super(type.getCanonicalName() + "." + name);
		itsProvider      = provider;
		itsType          = type;
		itsPredicate	 = predicate;
		itsObjects       = itsProvider.getEntitiesByType( itsType );
		itsResults       = null;
		itsIterator      = null;
		
		itsProvider.insertNamedQuery(type,this);
	}

	/************************************************************************
	 * Creates a new InMemoryNamedQuery. 
	 *
	 */
	public 
	InMemoryNamedQuery(InMemoryNamedQuery<T> other)
	{
		super(other);
		itsProvider       = other.itsProvider;
		itsType          = other.itsType;
		itsPredicate	 = other.itsPredicate;
		itsObjects       = itsProvider.getEntitiesByType( itsType );
		itsResults       = null;
		itsIterator      = null;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public InMemoryNamedQuery<T> 
	copy()
	{
	    return new InMemoryNamedQuery<T>( this );
	}

	/************************************************************************
     * {@inheritDoc} 
	 * @return 
     */
    @Override
    public INamedQuery<T> 
    clear()
    {
        itsObjects = itsProvider.getEntitiesByType( itsType );
        itsResults = null;
        itsIterator = null;
        return this;
    }

    /************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public T 
	getUnique() 
	    throws NotUniqueException
	{
		execute();
		
		if ( itsResults.size() > 1 )
		    throw new NotUniqueException("result is not unique");
		
		if ( itsResults.isEmpty() )
		    return null;
		
		return itsResults.get( 0 );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public Collection<T> 
	getAll()
	{
		execute();
		return itsResults;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public T 
	getNext()
	{
		execute();
		return itsIterator.next();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public boolean 
	hasNext()
	{
		execute();
		return itsIterator.hasNext();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
    public boolean 
    hasUnique() 
        throws NotUniqueException
    {
	    execute();
	    return itsResults.size() == 1;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    hasAny()
    {
        execute();
        return !itsResults.isEmpty();
    }

    /************************************************************************
	 *  
	 *
	 */
	protected void
	execute()
	{
		if ( itsResults != null )
			return;
		
		itsObjects  = itsProvider.getEntitiesByType( itsType );
		itsResults  = getResults( getInputs().getNamedInputs(),itsObjects );
		itsIterator = itsResults.iterator();
	}
	
	/************************************************************************
	 *  
	 *
	 * @param inputs
	 * @param objects
	 * @return
	 */
	protected List<T> 
	getResults(Map<String,Object> inputs,List<T> objects)
	{
		List<T> output = new ArrayList<T>();
		
		for (T entry : objects)
			if ( itsPredicate.evaluate( entry,inputs ) )
				output.add( entry );
		
		return output;
	}
}


// ##########################################################################
